### DISCUSSION TITLE
Should Porffor have node compat?

### DISCUSSION THREAD

**Discussion by CanadaHonk** [CanadaHonk/porffor#9] [General]
Commonly used Node-y globals like `process` (eg `process.stdout.write`, `process.versions.xxxxx`)

**Comment by kran6a** [comment]
While I would like the idea of compiling my node services to native code I don't think it is the compiler's responsability to support anything further than js/ts itself.

It could add support for addons so it could support different runtimes and non-standard features.
The main problem I see with adding support for Node is that people will (rightfully) ask for support for Deno, Bun and maybe WebStandards too.
It could be also argued that Porforr is its own "runtime" so it could just add its own features on top of the language which again I don't think it should do.

In my opinion addons should be invoked via import attributes. For example, this should import the corresponding file using the node addon.
```ts
import {process} from "node:process" with { adapter: "node" };
```
The node addon would be a separate package from porforr (i.e: @porforr/node-addon)


We could also add support for porforr-specific extensions the same way. For example, compile time macros would be a nice compiler feature since macros allow slimming down binaries as we skip adding their code and dependencies. We could access the time the binary was built and the hash of the package_lock.json used to build it with the following code.

```ts
// build_info.ts | macro
import {createHash} from "node:crypto";
import {readFileSync} from "node:fs";
export const build_date = new Date().toString();
export const package_lock_hash = createHash("sha256").update(readFileSync("./package_lock.json")).digest("hex");
```
```ts //index.ts
import {build_date, package_lock_hash} from "./build_info.js" with { adapter: "node", type: "macro" };
```

Importing non-js files would be another use-case for addons. ie:
```ts //index.ts
import background_image from "./assets/bg.webp" with { adapter: "image/webp", type: "macro", options: {quality: 80, size: {width: 1920, height: 1080}} }; // background_image is a Buffer containing a compressed and resized image generated by the image/webp addon
```

Macros would also enable interop between code meant for otherwise non-compatible runtimes.

**Comment by guest271314** [comment]
Just implement reading `stdin` and writing to `stdout`, including binary input/output support, which many JavaScript engines, runtimes, interpreters do, though not all do; and is actually outside of the scope of ECMA-262 because there's no I/O at all for the language per that specification. I would suggest to not get in to chasing Node.js API's.
